### ISSUE TITLE
IOS - VBA - Keyboard is overlapping the company website filed under Company Information page

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#9542]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Login in App
2. Go to Settings - Workspace - Connect bank account
3. Use credentials: user_good/pass_good
4. Select Wells Fargo and than Saving account
5. Start to enter data into Company Information page
6. In step 2 (company information) enter the phone number and start entering company website

## Expected Result:
Keyboard is not overlapping the company website filed under Company Information page

## Actual Result:
Keyboard is overlapping the company website filed under Company Information page

## Workaround:
Unknow

## Platform:

Where is this issue occurring?

- iOS

**Version Number:** 1.1.78.7

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**Email or phone of affected tester (no customers):** any

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/175130431-ea81108b-981b-4e05-91f1-8d99f930625b.mp4

**Expensify/Expensify Issue URL:** 

**Issue reported by:** Applause - Internal Team

**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by Christinadobrzyn** [comment]
I think this might be related to Accessibility - reaching out in open-source for a buddy check - https://expensify.slack.com/archives/C01GTK53T8Q/p1656108828757429

**Comment by Christinadobrzyn** [comment]
Nope sounds like a bug! Created Upwork job

Internal positing - https://www.upwork.com/ab/applicants/1541435223002427392/job-details
External positing - https://www.upwork.com/jobs/~01734e552bce49e505

**Comment by Karim-30** [comment]
### Proposal

`Phone number` text input opens numeric keyboard and `Company website` text input opens alphabetical keyboard, alphabetical keyboard is taller than numeric keyboard, so if we move `Company website` text input above `Phone number` text input we guarantee that switching from an upper text input to a lower text input will cause a keyboard shrinking. 

I know it is a very simple solution, but I tried to wrap the code inside `KeyboardAvoidingView` component and no luck, so rather than fix out `KeyboardAvoidingView` component or create a new wrapper component and use it only on one page (we don't have any page has this number of input fields ), we can use this simple solution.

Result :

https://user-images.githubusercontent.com/108357004/176237490-3f8983a3-3aeb-4947-aaae-892954a0fe53.mov

**Comment by mananjadhav** [comment]
@Karim-30 Thank you for the comment, but your solution looks like a workaround and doesn't solve the actual problem. Looking for a proposal that probably adjusts the view on focus of the input.

**Comment by mananjadhav** [comment]
Waiting for proposals.

**Comment by puneetlath** [comment]
Looks like we're still collecting proposals for this job - price increase to $1000

**Comment by puneetlath** [comment]
Still waiting for proposals. Upping to $2000.

**Comment by Karim-30** [comment]
### Proposal

In `src/components/FormScrollView.js` : 

```
  import styles from '../styles/styles';
+ import variables from '../styles/variables';

  ...

    <ScrollView
        style={[styles.w100, styles.flex1]}
        ref={ref}
        contentContainerStyle={styles.flexGrow1}
        keyboardShouldPersistTaps="handled"
+       contentInset={{bottom: variables.inputHeight}}
```

Resault : 

https://user-images.githubusercontent.com/108357004/179829185-5242ac41-de14-4b40-8827-f8e084e85716.mov

**Comment by mananjadhav** [comment]
@Karim-30 Thanks for your proposal, contentInset might work but it does impact even when the keyboard is not open.  I guess we don't want that. 

Meanwhile, I am going to check other impacts for putting this.

**Comment by Karim-30** [comment]
@mananjadhav Thank you for the review,  we can remove any impacts that happen when the keyboard is not open by using keyboard listeners `keyboardDidShow`, `keyboardDidHide`

**Comment by mananjadhav** [comment]
@Karim-30 Can yeah we can remove the impacts but can you update your proposal to handle the change please? Also a small suggestion we have `withKeyboardState` that uses `keyboardDidShow` & `keyboardDidHide`

**Comment by errahsoufiane** [comment]
@mananjadhav do you have any thought on my previous [proposal](https://github.com/Expensify/App/issues/9542#issuecomment-1189452636)?

### **Updated Proposal**
https://github.com/Expensify/App/blob/712ca243a4c0b4101fb8e3c9da5aeac7132e153c/src/pages/ReimbursementAccount/ReimbursementAccountForm.js#L15
Add this line:
```
import withKeyboardState from '../../components/withKeyboardState';
```


https://github.com/Expensify/App/blob/712ca243a4c0b4101fb8e3c9da5aeac7132e153c/src/pages/ReimbursementAccount/ReimbursementAccountForm.js#L74
Add this line:
```
    withKeyboardState,
```


https://github.com/Expensify/App/blob/712ca243a4c0b4101fb8e3c9da5aeac7132e153c/src/pages/ReimbursementAccount/ReimbursementAccountForm.js#L49
Add this line:
```
contentInset={{bottom: this.props.isShown ? 48 : 0}}
```


And this fix is not only for this issue bot throughout the app which show native keyboard for input:
https://github.com/Expensify/App/blob/712ca243a4c0b4101fb8e3c9da5aeac7132e153c/src/components/ScreenWrapper.js#L136
```
                            <KeyboardAvoidingView style={[styles.w100, styles.h100]} behavior={this.props.keyboardAvoidingViewBehavior} enabled={Platform.OS === 'ios'}>
```

Before fix

iOS:

https://user-images.githubusercontent.com/108141252/180840335-f5706288-6c10-4f6e-b2f9-1844165a19f7.mp4


Android:

https://user-images.githubusercontent.com/108141252/180840383-1aa01639-7d6e-4abc-a360-a8b75b61ab9d.mp4



After fix

iOS:

https://user-images.githubusercontent.com/108141252/180840479-57ea7ff7-1ee2-41f1-b358-4d072eadaa90.mp4


Android:

https://user-images.githubusercontent.com/108141252/180840558-e51c48b9-d8a0-48ef-a0d6-2ef94243a8dd.mp4

**Comment by mananjadhav** [comment]
@errahsoufiane I don't see any proposal from you that describes any implementation details. In the last comment it seems you're linking @Karim-30's proposal and extending to what they have already figured out. 

Thank you for taking time, but I will hear from @Karim-30's updated proposal here.

**Comment by errahsoufiane** [comment]
@mananjadhav I had proposed general approach and once you or any other reviewer replied, I was going to share detailed implementation with code sharing.
BTW I don't mind whose proposal is accepted for this issue (just to make project goes well) but 2 more suggestions from my updated proposal:
- No need to apply contentInset for all scroll views (FormScrollView) throughout the app. The issue happens only on this screen, therefore it's enough to apply this only for ReimbursementAccountForm
- While fixing this issue, you may need to fix for other platforms (especially Android) as well. As you see my videos attached, it's very weird behavior in android everywhere keyboard showing throughout the app. I also proposed this fix. (this might fix other open issues related to keyboard behavior)

**Comment by Karim-30** [comment]
### Updated Proposal 

In src/components/FormScrollView.js :

```
  import styles from '../styles/styles';
+ import variables from '../styles/variables';
+ import withKeyboardState from './withKeyboardState';

  ...

    <ScrollView
        style={[styles.w100, styles.flex1]}
        ref={ref}
        contentContainerStyle={styles.flexGrow1}
        keyboardShouldPersistTaps="handled"
+       contentInset={props.isShown ? {bottom: variables.componentSizeLarge / 2} : undefined}

```
Result :

https://user-images.githubusercontent.com/108357004/180860550-97308d97-0c14-4caa-9609-3a564e45f72e.mov

**Comment by errahsoufiane** [comment]
@mananjadhav I want you to go ahead with @Karim-30 's proposal but I found one issue:
`variables.componentSizeLarge / 2` this padding is not enough when auto suggestions show:

https://user-images.githubusercontent.com/108141252/180862104-abd6f0a4-1211-4381-8cfd-4298a8946dbc.MP4

**Comment by Karim-30** [comment]
@errahsoufiane Thanks for testing.

### Updated Proposal
In src/components/FormScrollView.js :
```
  import styles from '../styles/styles';
+ import variables from '../styles/variables';
+ import withKeyboardState from './withKeyboardState';

  ...

    <ScrollView
        style={[styles.w100, styles.flex1]}
        ref={ref}
        contentContainerStyle={styles.flexGrow1}
        keyboardShouldPersistTaps="handled"
+       contentInset={props.isShown ? {bottom: variables.componentSizeLarge} : undefined}
```

Result : 

https://user-images.githubusercontent.com/108357004/180868954-4853a4f5-40ef-4ca7-83b4-770b6456455a.mov


There is a gap below the button after closing the keyboard, it hides after making a scroll, I tried to remove this gap by `bottom: variables.componentSizeLarge / 2` but it seems not working on small screen devices, So there is noway from hiding this gap.

**Comment by b1tjoy** [comment]
### PROPOSAL
### Problem
When keyboard change it's size, ScrollView's content does not adjust accordingly, so the content get blocked. After some digging, I find out that this issue not only occur in App develop with react-native, but also can occur in iOS native App, it's caused by iOS native code. Fix it in react-native library is a better choice.

### Solution
In react-native 0.68.0, ScrollView has a new prop `automaticallyAdjustKeyboardInsets` for iOS, which controls whether the ScrollView should automatically adjust its contentInset and scrollViewInsets when the Keyboard changes its size.

Here is the solution:

1. Upgrade react-native to 0.68.0 or higher version in our project
2. Add prop `automaticallyAdjustKeyboardInsets` in [FormScrollView.js](https://github.com/Expensify/App/blob/main/src/components/FormScrollView.js)

https://github.com/Expensify/App/blob/a06358cef3ab1951fcbb5c256c003acdf238d447/src/components/FormScrollView.js#L12-L16
```JSX
+  automaticallyAdjustKeyboardInsets
```
https://github.com/Expensify/App/blob/a06358cef3ab1951fcbb5c256c003acdf238d447/src/components/FormScrollView.js#L17-L21

### Screenshot
Since upgrade react-native to 0.68.0 or higher needs lots of work, I write a minimum code project to demonstrate how prop 'automaticallyAdjustKeyboardInsets` make ScrollView automatically adjust to prevent its content get blocked when keyboard changes its size. Will upload the video later.

**Comment by b1tjoy** [comment]
### UPDATE
The previous proposal https://github.com/Expensify/App/issues/9542#issuecomment-1200443420 has some problem, text input may be pushed out of sight in some test scenario, we can't use prop `automaticallyAdjustKeyboardInsets` to fix current issue.

### PROPOSAL
### Problem
When keyboard change it's size, ScrollView's content does not adjust accordingly, so the content get blocked. We can calculate how much the text input is blocked by keyboard, then manually scroll content view to appropriate position, and make sure text input is in visible area.

### Solution

1. Make a new directory name `FormScrollView` in `src/components`
2. Move `src/components/FormScrollView.js` to `src/components/FormScrollView/index.js`, fix import path error
3. Touch a new file `src/components/FormScrollView/index.ios.js`, and add the following code to it

```JSX
import React from 'react';
import PropTypes from 'prop-types';
import {Keyboard, ScrollView, TextInput} from 'react-native';
import styles from '../../styles/styles';
import _ from 'underscore';

const propTypes = {
    /** Form elements */
    children: PropTypes.node.isRequired,

    /** A ref to forward to the scroll view */
    forwardedRef: PropTypes.func,
};

const defaultProps = {
    forwardRef: null,
};

class FormScrollView extends React.Component {
    constructor(props) {
        super(props);
        this.keyboardHeight = React.createRef(0);
        this.scrollPosition = React.createRef(0);
        this.keyboardDidChangeFrameListener = this.keyboardDidChangeFrameListener.bind(this);
    }

    componentDidMount() {
        this.keyboardDidChangeFrameSubscription = Keyboard.addListener('keyboardDidChangeFrame', this.keyboardDidChangeFrameListener);
    }

    componentWillUnmount() {
        if (this.keyboardDidChangeFrameSubscription) {
            this.keyboardDidChangeFrameSubscription.remove();
        }
    }

    keyboardDidChangeFrameListener(keyboardFrame) {
        const oldKeyboardHeight = this.keyboardHeight.current;
        this.keyboardHeight.current = keyboardFrame.endCoordinates.height;
        if (oldKeyboardHeight === keyboardFrame.endCoordinates.height) {
            // Keyboard size does not change, just return
            return;
        }

        const currentlyFocusedRef = TextInput.State.currentlyFocusedInput();
        if (currentlyFocusedRef) {
            currentlyFocusedRef.measureInWindow((x, y, width, height) => {
                const textInputBottomPosition = y + height;
                const keyboardPosition = keyboardFrame.endCoordinates.screenY;
                if (textInputBottomPosition > keyboardPosition) {
                    // The text input get blocked, manually scroll to the right position
                    const scrollPosition = this.scrollPosition.current + textInputBottomPosition - keyboardPosition;
                    this.form.scrollTo({y: scrollPosition, animated: true});
                }
            });
        }
    }

    render() {
        return (
            <ScrollView
                style={[styles.w100, styles.flex1]}
                ref={(ref) => {
                    this.form = ref;
                    if (_.isFunction(this.props.forwardRef)) {
                        this.props.forwardRef(ref);
                    }
                }}
                onScroll={(event) => {
                    this.scrollPosition.current = event.nativeEvent.contentOffset.y;
                }}
                scrollEventThrottle={50}
                contentContainerStyle={styles.flexGrow1}
                keyboardShouldPersistTaps="handled"
                // eslint-disable-next-line react/jsx-props-no-spreading
                {...this.props}
            >
                {this.props.children}
            </ScrollView>
        );
    }
}

FormScrollView.propTypes = propTypes;
FormScrollView.defaultProps = defaultProps;

export default React.forwardRef((props, ref) => (
    <FormScrollView {...props} forwardRef={ref} />
));
```

What do we do in iOS specific `FormScrollView` component?

1. Add props scrollEventThrottle and onScroll event handler to track current scroll position
2. Add keyboardDidChangeFrame event handler to track keyboard change frame event
3. Check if keyboard size changes and text input is blocked by keyboard in event handler
4. Calculate how much the text input is blocked by keyboard and scroll to right position

### Screenshot

https://user-images.githubusercontent.com/103875612/182357319-cf89665f-f434-4ca5-bbc1-5c487f2717bb.mp4

**Comment by mananjadhav** [comment]
Please allow me a day, I am OOO and will review tomorrow.

**Comment by mananjadhav** [comment]
Reviewing @b1tjoy's proposal.

@Karim-30 I don't like the idea of adding hardcode trial and error fixes. This will end up breaking some or the other device resolution.

@errahsoufiane This is still open for review so if your proposal is any different from the existing ones, feel free to post it.

**Comment by puneetlath** [comment]
Increasing price to $4000.

**Comment by mananjadhav** [comment]
@b1tjoy I went through your proposal and also explored other alternatives meanwhile, I have a question, could you please explain in which cases is the text input going out of section when used with `automaticallyAdjustKeyboardInsets`. If say some issue does exist with `automaticallyAdjustKeyboardInsets` it would make sense to fix that rather than have a workaround for us to track the scroll with keyboard events. Let me know what you think? I checked thread and the PR for `automaticallyAdjustKeyboardInsets` and it seems it was intended to solve this current issue only! We can explain manually tracking the scroll as a secondary option.

**Comment by b1tjoy** [comment]
@mananjadhav When I submit https://github.com/Expensify/App/issues/9542#issuecomment-1200443420, we use react-native 0.66.4 in our App, I backport the PR for `automaticallyAdjustKeyboardInsets` to react-native 0.66.4, and I found out that the text input go out of section in some case.

Our current main branch use react-native 0.69.4, I retest with `automaticallyAdjustKeyboardInsets`, the text input won't goes out of section, but it still cause other issue. When we use `ScrollView` with `automaticallyAdjustKeyboardInsets`, it provide keyboard avoiding feature like `KeyboardAvoidingView`. If we use `KeyboardAvoidingView` wrap around `ScrollView`, the scrollable area will become smaller than we expected if keyboard pop out.

Check out the following screen recording, pay attention to the scroll bar on the right hand.

https://user-images.githubusercontent.com/103875612/187058097-5cecb229-20e6-4312-bc6f-99c4e8d84752.mp4

If we do not use `KeyboardAvoidingView` wrap around `ScrollView`, the scrollalbe area won't get smaller, and the company website field won't blocked by keyboard. Here is a simple modification to show it really works. If we use this, the final code will be platform specific, more complex and not affect other pages and components.

1. Disable `KeyboardAvoidingView` in ScreenWrapper.js

https://github.com/Expensify/App/blob/e21736d87cc1ed95c1f64a54040ce242ced7ebc0/src/components/ScreenWrapper/BaseScreenWrapper.js#L80

```JSX
- <KeyboardAvoidingView style={[styles.w100, styles.h100]} behavior={this.props.keyboardAvoidingViewBehavior}>
+ <KeyboardAvoidingView enable={false} style={[styles.w100, styles.h100]} behavior={this.props.keyboardAvoidingViewBehavior}>
```

2. Add `automaticallyAdjustKeyboardInsets` to FormScrollView.js

https://github.com/Expensify/App/blob/e21736d87cc1ed95c1f64a54040ce242ced7ebc0/src/components/FormScrollView.js#L12-L16
```JSX
+ automaticallyAdjustKeyboardInsets
```

Screenshot for the modification above.


https://user-images.githubusercontent.com/103875612/187058463-a12afde5-3e5a-4e48-83c0-e64bbdd6bc45.mp4

I also notice that some text input is pushed out of section first and then pulled into second, that is annoying. I will check the PR which brings `automaticallyAdjustKeyboardInsets` into `ScrollView` and find out is there anything I can do with it.


https://user-images.githubusercontent.com/103875612/187058693-e8ddccf4-5b6b-4d22-ae9d-925f32758a1b.mp4

**Comment by mananjadhav** [comment]
I'll check your findings @b1tjoy. Thanks for the post.

**Comment by mananjadhav** [comment]
Thanks @b1tjoy for the detailed explanation. I really appreciate the effort you've put into the findings.

I agree that this can't be fixed in `react-native`, as I checked it myself. I am going to review your comments again and give you a final go ahead by this weekend/monday. Thank you for the patience on this one.

**Comment by mananjadhav** [comment]
@b1tjoy Your proposal looks good to me https://github.com/Expensify/App/issues/9542#issuecomment-1235749355. Any kinks can be handled in the PR review while testing.

I was a bit concerned about how reliable is this `const currentlyFocusedRef = TextInput.State.currentlyFocusedInput();`  But it seems it was replaced from `currentlyFocusedField` and does work as expected.

@puneetlath All yours.

🎀 👀 🎀 
C+ reviewed

**Comment by mallenexpensify** [comment]
We're putting all keyboard-related issues on hold for a bit while we discuss a more holistic approach to fixing them (also [this ](https://github.com/Expensify/App/issues/9620#top)and [this one](https://github.com/Expensify/App/issues/9252)). 

@b1tjoy it's clear you've invested a lot of time into this issue, your proposal is extremely thorough and well-outlined.  You will be compensated for your work, before disscussing the amount, we want to see if the best path forward is for you to complete the issue.  If so, we'd take off hold then you'd finish then get compensated 7 days after the PR hits production, assuming no regressions.  We're discussing next steps internally and will have another update on the hold timeline soon (hopefully tomorrow)

**Comment by puneetlath** [comment]
@b1tjoy @mananjadhav we didn't end up going with the proposed solution, but we would like to pay you a partial bounty for the time put into this issue. Can you both please apply to this job? 

https://www.upwork.com/jobs/~01ce8cc4db0296323e

**Comment by mananjadhav** [comment]
Thanks for this @puneetlath. Applied.

**Comment by puneetlath** [comment]
All paid, thanks!
